File: src/main/java/com/server/ordering/service/OwnerService.java
package com.server.ordering.service;

import com.server.ordering.domain.MemberType;
import com.server.ordering.domain.PhoneNumber;
import com.server.ordering.domain.dto.request.OwnerSignUpDto;
import com.server.ordering.domain.dto.request.PasswordChangeDto;
import com.server.ordering.domain.dto.request.SignInDto;
import com.server.ordering.domain.member.Owner;
import com.server.ordering.repository.OwnerRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import javax.persistence.NoResultException;
import javax.persistence.PersistenceException;
import java.util.List;
import java.util.Objects;
import java.util.Optional;

@Service
@Transactional(readOnly = true)
@RequiredArgsConstructor
@Slf4j
public class OwnerService implements MemberService<Owner> {

    private final OwnerRepository ownerRepository;
    private final PasswordEncoder passwordEncoder;

    /**
     * 점주 회원가입
     */
    @Transactional
    @Override
    public Long signUp(Object signUpDto) throws PersistenceException {
        OwnerSignUpDto ownerSignUpDto = (OwnerSignUpDto) signUpDto;
        PhoneNumber phoneNumber = new PhoneNumber(ownerSignUpDto.getPhoneNumber(), MemberType.OWNER);
        String encodedPassword = passwordEncoder.encode(ownerSignUpDto.getPassword());
        Owner owner = new Owner(ownerSignUpDto.getSignInId(), encodedPassword, phoneNumber);
        ownerRepository.save(owner);
        return owner.getId();
    }

    /**
     * 점주 로그인
     */
    @Transactional
    @Override
    public Owner signIn(SignInDto signInDto) throws PersistenceException {
        try {
            Owner owner = ownerRepository.findOneWithRestaurantById(signInDto.getSignInId());
            boolean bMatch = passwordEncoder.matches(signInDto.getPassword(), owner.getPassword());

            if (!bMatch) {
                return null;
            }

            owner.getRestaurant().putFirebaseToken(signInDto.getFirebaseToken());
            return owner;
        } catch (NoResultException e) {
            return null;
        }
    }

    @Transactional
    @Override
    public void signOut(Long id) {
        Owner owner = ownerRepository.findOneWithRestaurant(id);
        owner.getRestaurant().clearFirebaseToken();
    }


    /**
     * 점주 회원가입할 때 email 중복 검증
     */
    @Override
    public boolean isIdDuplicated(String signInId) throws PersistenceException {
        try {
            ownerRepository.findById(signInId);
            return true;
        } catch (NoResultException e) {
            return false;
        }
    }

    /**
     * 점주 휴대폰번호 변경
     */
    @Transactional
    public void putPhoneNumber(Long ownerId, String phoneNumber) {
        Owner owner = ownerRepository.findOneWithPhoneNumber(ownerId);
        owner.getPhoneNumber().putPhoneNumber(phoneNumber);
    }

    /**
     * 점주 비밀번호 변경
     */
    @Transactional
    public Boolean putPassword(Long ownerId, PasswordChangeDto dto) {
        Owner owner = ownerRepository.findOne(ownerId);
        if (passwordEncoder.matches(dto.getCurrentPassword(), owner.getPassword())) {
            String encodedNewPassword = passwordEncoder.encode(dto.getNewPassword());
            owner.putPassword(encodedNewPassword);
            return true;
        }
        return false;
    }

    /**
     * 점주 회원탈퇴
     */
    @Transactional
    @Override
    public void deleteAccount(Long id) {
        ownerRepository.remove(id);
    }
}

File: src/main/kotlin/com/byeoru/ordering_server/service/OwnerService.kt
package com.byeoru.ordering_server.service

import com.byeoru.ordering_server.domain.MemberType
import com.byeoru.ordering_server.domain.PhoneNumber
import com.byeoru.ordering_server.domain.dto.request.OwnerSignUpDto
import com.byeoru.ordering_server.domain.dto.request.PasswordChangeDto
import com.byeoru.ordering_server.domain.dto.request.SignInDto
import com.byeoru.ordering_server.domain.member.Owner
import com.byeoru.ordering_server.repository.OwnerRepository
import org.slf4j.LoggerFactory
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import javax.persistence.NoResultException
import javax.persistence.PersistenceException

@Service
@Transactional(readOnly = true)
class OwnerService(private val ownerRepository: OwnerRepository,
                   private val passwordEncoder: PasswordEncoder) : MemberService<Owner> {

    /**
     * 점주 회원가입
     */
    @Transactional
    @Throws(PersistenceException::class)
    override fun signUp(signUpDto: Any): Long? {
        val ownerSignUpDto = signUpDto as OwnerSignUpDto
        val phoneNumber = PhoneNumber(ownerSignUpDto.phoneNumber, MemberType.OWNER)
        val encodedPassword = passwordEncoder.encode(ownerSignUpDto.password)
        val owner = Owner(ownerSignUpDto.signInId, encodedPassword, phoneNumber)
        ownerRepository.save(owner)
        return owner.id
    }

    /**
     * 점주 로그인
     */
    @Transactional
    override fun signIn(signInDto: SignInDto): Owner? {
        return try {
            val owner = ownerRepository.findOneWithRestaurantById(signInDto.signInId)
            val bMatch = passwordEncoder.matches(signInDto.password, owner.password)
            if (!bMatch) {
                return null
            }
            owner.putFirebaseToken(signInDto.firebaseToken)
            owner
        } catch (e: NoResultException) {
            null
        }
    }

    @Transactional
    override fun signOut(id: Long) {
        val owner = ownerRepository.findOne(id)
        owner.clearFirebaseToken()
    }

    /**
     * 점주 회원가입할 때 email 중복 검증
     */
    @Throws(PersistenceException::class)
    override fun isIdDuplicated(signInId: String): Boolean {
        return try {
            ownerRepository.findById(signInId)
            true
        } catch (e: NoResultException) {
            false
        }
    }

    /**
     * 점주 휴대폰번호 변경
     */
    @Transactional
    fun putPhoneNumber(ownerId: Long, phoneNumber: String) {
        val owner = ownerRepository.findOneWithPhoneNumber(ownerId)
        owner.phoneNumber.putPhoneNumber(phoneNumber)
    }

    /**
     * 점주 비밀번호 변경
     */
    @Transactional
    fun putPassword(ownerId: Long, dto: PasswordChangeDto): Boolean {
        val owner = ownerRepository.findOne(ownerId)
        if (passwordEncoder.matches(dto.currentPassword, owner.password)) {
            val encodedNewPassword = passwordEncoder.encode(dto.newPassword)
            owner.putPassword(encodedNewPassword)
            return true
        }
        return false
    }

    /**
     * 점주 회원탈퇴
     */
    @Transactional
    override fun deleteAccount(id: Long) = ownerRepository.remove(id)

    companion object {
        private val log = LoggerFactory.getLogger(OwnerService::class.java)
    }
}